README.org
#+UUID: f8c991e5-f6c7-3b6b-2547-1e344d751bc1

A library for pretty-printing s-expressions, using better indentation rules than
the default pretty printer in Sexplib.


CHANGES.md
## Release v0.16.0

* Added `Bright` versions of colors to `Config.color`.

* Refer to `Stdlib` instead of `Caml`.


test/import.ml
include Core_kernel

module Pretty_print_config = Sexp_pretty.Pretty_print_config


test/test_comments.ml
(** We have two kinds of tests:

    1. Handwritten tests showing the pretty strings for some known examples.

    2. Quickcheck tests which round-trip a sexp through pretty-printing and assert that
    the comments are preserved up to some post-processing of whitespace.

    The post-processing is applied to both the original comments and the pretty comments.
    It describes how closely we can look before we see how pretty-printing has changed
    comments.
*)

open! Import
open! Expect_test_helpers_core
open! Sexp_pretty

module Comment_print_style = struct
  type t = Config.comment_print_style =
    | Pretty_print
    | Conservative_print
  [@@deriving enumerate, sexp_of]

  let to_string t = Sexp.to_string_mach ([%sexp_of: t] t)
end

type t = Sexp.With_layout.t_or_comment list [@@deriving sexp_of]

let get_config ~comment_print_style =
  Config.update
    ~color:false
    { Config.default with
      comments =
        (match Config.default.comments with
         | Drop -> raise_s [%message "Expected default to be Print"]
         | Print (comment_indent, color_option, _) ->
           Print (comment_indent, color_option, comment_print_style))
    }
;;

let of_string s : t =
  match
    Lexing.from_string s
    |> Sexplib.Sexp.With_layout.Parser.sexps Sexplib.Sexp.With_layout.Lexer.main
  with
  | exception exn -> raise_s [%message "sexps_with_layout_of_string" (exn : exn) s]
  | t -> t
;;

let pretty_string comment_print_style list =
  let config = get_config ~comment_print_style in
  List.map list ~f:(Sexp_with_layout.pretty_string config) |> String.concat ~sep:" "
;;

let%expect_test "leading and trailing whitespace" =
  let module Example = struct
    type t =
      { leading : bool
      ; trailing : bool
      }
    [@@deriving enumerate, sexp_of]

    let to_string { leading; trailing } =
      let space_if bool = if bool then " " else "" in
      [%string {|#|%{space_if leading}a%{space_if trailing}|#|}]
    ;;

    let to_string_pretty t ~style =
      t
      |> to_string
      |> of_string
      |> pretty_string style
      (* Visualize the whitespace *)
      |> String.tr ~target:' ' ~replacement:'.'
    ;;

    let columns =
      Ascii_table_kernel.Column.create "comment" to_string
      :: List.map [%all: Comment_print_style.t] ~f:(fun style ->
        Ascii_table_kernel.Column.create
          (Comment_print_style.to_string style)
          (to_string_pretty ~style))
    ;;
  end
  in
  let ascii_table columns rows =
    let screen =
      Ascii_table_kernel.draw columns rows ~prefer_split_on_spaces:false
      |> Option.value_exn ~here:[%here]
    in
    Ascii_table_kernel.Screen.to_string
      screen
      ~bars:`Unicode
      ~string_with_attr:(fun _ s -> s)
  in
  ascii_table Example.columns [%all: Example.t] |> print_endline;
  [%expect
    {|
    ┌─────────┬──────────────┬────────────────────┐
    │ comment │ Pretty_print │ Conservative_print │
    ├─────────┼──────────────┼────────────────────┤
    │ #|a|#   │ #|.a.|#      │ #|a|#              │
    │ #| a|#  │ #|.a.|#      │ #|.a|#             │
    │ #|a |#  │ #|.a.|#      │ #|a.|#             │
    │ #| a |# │ #|.a.|#      │ #|.a.|#            │
    └─────────┴──────────────┴────────────────────┘ |}]
;;

let%expect_test "examples" =
  let test s =
    let t = of_string s in
    List.iter [%all: Comment_print_style.t] ~f:(fun style ->
      let s = pretty_string style t in
      print_endline [%string {|%{style#Comment_print_style}: %{s}|}])
  in
  (* multiple internal whitespace *)
  test {| #| a  b |# |};
  [%expect {|
    Pretty_print: #| a b |#

    Conservative_print: #| a  b |# |}];
  (* multiple lines *)
  test {| #| a
b
 c

   d |# |};
  [%expect
    {|
    Pretty_print: #| a b c d |#

    Conservative_print: #| a
    b
     c

       d |# |}];
  (* really long comment *)
  test
    {| #| really long comment to see if sexp pretty printing will ever add new line breaks |# |};
  [%expect
    {|
    Pretty_print: #|
       really long comment to see if sexp pretty printing will ever add new line
       breaks
    |#

    Conservative_print: #| really long comment to see if sexp pretty printing will ever add new line breaks |# |}];
  (* comments nested in sexp comment *)
  test
    {| #;
    ; plain comment
    #| block comment |#
    (sexp
with
#| nested comment |#
) |};
  [%expect
    {|
    Pretty_print: #;
    ; plain comment#| block comment |#
    (sexp with #| nested comment |#)

    Conservative_print: #;
    ; plain comment#| block comment |#
    (sexp with #| nested comment |#) |}];
  (* quoted newline *)
  test {| #| "quoted
newline" |# |};
  [%expect
    {|
    Pretty_print: #| "quoted newline" |#

    Conservative_print: #| "quoted
    newline" |# |}];
  ignore ()
;;

let%expect_test "sticky comments" =
  let test s =
    let t = of_string s in
    Config.[ Before; After; Same_line ]
    |> List.iter ~f:(fun sticky ->
      let label = [%sexp_of: Config.sticky_comments] sticky |> Sexp.to_string in
      print_endline [%string {|%{label}:|}];
      List.iter [%all: Comment_print_style.t] ~f:(fun comment_print_style ->
        let s =
          let config = get_config ~comment_print_style in
          let config = { config with Config.sticky_comments = sticky } in
          List.map t ~f:(Sexp_with_layout.pretty_string config)
          |> String.concat ~sep:" "
        in
        print_endline [%string "%{comment_print_style#Comment_print_style}:\n%{s}"]))
  in
  test
    {|
 (a
    ;; comment 1
    ;; comment 2

    (b c) ;; comment 3
    ( ;; comment 4
      d ;; comment 5

    #| multi line
    block
    comment
    |#
    ()
    ))
|};
  [%expect
    {|
    Before:
    Pretty_print:
    (a
      ;; comment 1
      ;; comment 2
      ;; comment 3
      (b c)
      (;; comment 4
       ;; comment 5
       d
       #| multi line block comment |#
       ()))

    Conservative_print:
    (a
      ;; comment 1
      ;; comment 2
      ;; comment 3
      (b c)
      (;; comment 4
       ;; comment 5
       d
       #| multi line
        block
        comment
        |#
    ()))

    After:
    Pretty_print:
    (a
      ;; comment 1
      ;; comment 2
      (b c)
      ;; comment 3
      (;; comment 4
       d
       ;; comment 5
       #| multi line block comment |#
       ()))

    Conservative_print:
    (a
      ;; comment 1
      ;; comment 2
      (b c)
      ;; comment 3
      (;; comment 4
       d
       ;; comment 5
       #| multi line
        block
        comment
        |#
    ()))

    Same_line:
    Pretty_print:
    (a
      ;; comment 1
      ;; comment 2
      (b c) ;; comment 3
      (;; comment 4
       d ;; comment 5
       #| multi line block comment |#
       ()))

    Conservative_print:
    (a
      ;; comment 1
      ;; comment 2
      (b c) ;; comment 3
      (;; comment 4
       d ;; comment 5
       #| multi line
        block
        comment
        |#
    ()))
      |}];
  ignore ()
;;

let round_trip_pretty style tc = of_string (pretty_string style tc)

let get_comment_strings t =
  let open Sexp.With_layout in
  let rec of_t = function
    | Atom _ -> []
    | List (_, t_or_comments, _) -> List.concat_map t_or_comments ~f:of_t_or_comment
  and of_t_or_comment = function
    | Sexp _ -> []
    | Comment comment -> of_comment comment
  and of_comment = function
    | Plain_comment (_, s) -> [ s ]
    | Sexp_comment (_, comments, t) ->
      List.concat [ List.concat_map comments ~f:of_comment; of_t t ]
  in
  List.concat_map t ~f:of_t_or_comment
;;

let whitespace_chars = String.to_list " \t\n\r"

let compress_whitespace s =
  String.split_on_chars s ~on:whitespace_chars
  |> List.filter ~f:(Fn.non String.is_empty)
  |> String.concat ~sep:" "
;;

(* Doesn't strip sexp comments. Our comparison is on the [Plain_comment _] values,
   including those nested inside sexp comments. *)
let strip_comment_markers s =
  match String.chop_prefix s ~prefix:";" with
  | Some line_comment -> line_comment
  | None ->
    (match
       String.chop_prefix s ~prefix:"#|"
       |> Option.bind ~f:(String.chop_suffix ~suffix:"|#")
     with
     | Some s -> s
     | None -> raise_s [%message "Comment neither line comment nor block comment." s])
;;

let postprocess_comment (style : Comment_print_style.t) prev =
  match style with
  (* [Pretty_print] preserves non-whitespace characters *)
  | Pretty_print -> prev |> strip_comment_markers |> String.strip |> compress_whitespace
  (* [Conservative_print] preserves everything *)
  | Conservative_print -> prev
;;

module Test_helper = struct
  module type S = sig
    type t [@@deriving quickcheck, sexp_of]
  end

  type 'a t = (module S with type t = 'a)

  let filter_map (type a) ((module M) : a t) ~f : a t =
    (module struct
      type t = M.t [@@deriving sexp_of]

      include
        Quickcheckable.Of_quickcheckable_filtered
          (M)
          (struct
            type t = M.t

            let of_quickcheckable a = f a
            let to_quickcheckable = Fn.id
          end)
    end)
  ;;
end

(* [Sexp_pretty] doesn't promise to round-trip. Let's limit ourselves to inputs
   that round-tripped up to the number of comments. *)
let round_trippable_sexps style =
  let equal_num_comments a b =
    Comparable.lift
      [%compare.equal: int]
      ~f:(fun list -> get_comment_strings list |> List.length)
      a
      b
  in
  Test_helper.filter_map
    (module Sexp_string_quickcheck.Sexp_string)
    ~f:(fun s ->
      Option.try_with (fun () ->
        let prev = of_string s in
        let next = round_trip_pretty style prev in
        assert (equal_num_comments prev next);
        s))
;;

let test_one style =
  let equal a b =
    Comparable.lift
      [%equal: string list]
      ~f:(fun list -> get_comment_strings list |> List.map ~f:(postprocess_comment style))
      a
      b
  in
  stage (fun s ->
    let prev = of_string s in
    let next = round_trip_pretty style prev in
    if not (equal prev next)
    then raise_s [%message "Comments changed too much." s (pretty_string style next)])
;;

let%expect_test "quickcheck" =
  let test here style =
    require_does_not_raise here (fun () ->
      let module M = (val round_trippable_sexps style) in
      Base_quickcheck.Test.run_exn (module M) ~f:(unstage (test_one style)))
  in
  test [%here] Pretty_print;
  [%expect {| |}];
  test [%here] Conservative_print;
  [%expect {| |}];
  ignore ()
;;


src/sexp_pretty_intf.ml
open! Base

module type S = sig
  type sexp
  type 'a writer = Config.t -> 'a -> sexp -> unit

  (** [pp_formatter conf fmt sexp] will mutate the fmt with functions such as
      [set_formatter_tag_functions] *)
  val pp_formatter : Stdlib.Format.formatter writer

  val pp_formatter'
    :  next:(unit -> sexp option)
    -> Config.t
    -> Stdlib.Format.formatter
    -> unit

  val pp_buffer : Buffer.t writer
  val pp_out_channel : Stdlib.out_channel writer
  val pp_blit : (string, unit) Blit.sub writer

  (** [pretty_string] needs to allocate. If you care about performance, using one of the
      [pp_*] functions above is advised. *)
  val pretty_string : Config.t -> sexp -> string

  val sexp_to_string : sexp -> string
end

(** Pretty-printing of S-expressions *)
module type Sexp_pretty = sig
  module Config = Config

  module type S = S

  include S with type sexp := Sexp.t
  module Sexp_with_layout : S with type sexp := Sexplib.Sexp.With_layout.t_or_comment

  module Normalize : sig
    type t =
      (* Contains a sexp with associated comments. *)
      | Sexp of sexp * string list
      | Comment of comment

    and comment =
      | Line_comment of string
      (* Does not contain the "#|" "|#"; contains its indentation size. *)
      | Block_comment of int * string list
      | Sexp_comment of comment list * sexp

    and sexp =
      | Atom of string
      | List of t list

    val of_sexp_or_comment : Config.t -> Sexplib.Sexp.With_layout.t_or_comment -> t
  end

  val sexp_to_sexp_or_comment : Sexp.t -> Sexplib.Sexp.With_layout.t_or_comment
end


src/pretty_print_config.ml
open! Import


type color =
  | Black
  | Red
  | Green
  | Yellow
  | Blue
  | Magenta
  | Cyan
  | White
  | Default
[@@deriving sexp]

(* Datatypes of various thresholds *)
type atom_threshold = Atom_threshold of int [@@deriving sexp]
let atom_threshold_of_sexp sexp =
  match atom_threshold_of_sexp sexp with
  | Atom_threshold n when n<0 ->
    of_sexp_error "Atom threshold must be non_negative." sexp
  | threshold                 -> threshold
type char_threshold = Character_threshold of int [@@deriving sexp]
let char_threshold_of_sexp sexp =
  match char_threshold_of_sexp sexp with
  | Character_threshold n when n<0 ->
    of_sexp_error "Character threshold must be non_negative." sexp
  | threshold                      -> threshold
(* Depth is the depth of an atom. For example, in (a (b (c) d)), the depth of a is 1, the
   depth of b and d is 2, and depth of c is 3.
   Depth_threshold usually refers to the maximum depth of any atom in a list for it to be
   considered for certain heuristic, e.g. data alignment.
*)
type depth_threshold = Depth_threshold of int [@@deriving sexp]
let depth_threshold_of_sexp sexp =
  match depth_threshold_of_sexp sexp with
  | Depth_threshold n when n<1 ->
    of_sexp_error "Depth threshold must be greater than 0." sexp
  | threshold                      -> threshold

(* Whether or not should closing parentheses be aligned. *)
type aligned_parens =
  | Parens_alignment of bool
[@@deriving sexp]

type data_alignment =
  | Data_not_aligned
  (* Character threshold is excluding spaces and parentheses, the maximum depth can't exceed
     the depth threshold.
  *)
  | Data_aligned of aligned_parens * atom_threshold * char_threshold * depth_threshold
[@@deriving sexp]

type atom_coloring =
  (* Color the first one, only if the number of atoms that follow it at most the value of
     the constructor's argument.
  *)
  | Color_first of int
  | Color_all
  | Color_none
[@@deriving sexp]
let atom_coloring_of_sexp sexp =
  match atom_coloring_of_sexp sexp with
  | Color_first n when n<0 ->
    of_sexp_error "The limit to color atoms must be non-negative." sexp
  | coloring               -> coloring

(* This currently relates only to block comments. [Auto_indent] tries to infer the
   indentation from the original formatting, [Indent_comment n] indents new lines in a
   block comment by n spaces.
*)
type comment_indent =
  | Auto_indent_comment
  | Indent_comment of int
[@@deriving sexp]
let comment_indent_of_sexp sexp =
  match comment_indent_of_sexp sexp with
  | Indent_comment n when n<0 -> of_sexp_error "Indentation must be non-negative." sexp
  | indent                    -> indent

type comment_print_style =
  (* Auto aligns multi-line block comments. *)
  | Pretty_print
  (* Leaves block comments as they are, only adjusts indentation. *)
  | Conservative_print
[@@deriving sexp]

(* Comment treatment. *)
type comments =
  | Drop
  | Print of comment_indent * color option * comment_print_style
[@@deriving sexp]

type atom_printing =
  | Escaped     (* Can be parsed again. Atoms are printed out as loaded, with escaping. *)
  | Interpreted (* Try to interpret atoms as sexps. *)
[@@deriving sexp]

(* Singleton_lists are lists of the following format

   (ATOM_1 .. ATOM_N (....))

   and are printed in the following way if they are too big to fit on a line/force a
   breakline for other reasons:

   (ATOM_1 .. ATOM_N (
   ....
   ))

   Thresholds correspond to what's an acceptable number/size of the leading atoms ATOM_1
   through ATOM_N.

   Character threshold is excluding spaces.

*)
type singleton_limit =
  | Singleton_limit of atom_threshold * char_threshold [@@deriving sexp]

(* Should parentheses be colored? *)
type paren_coloring = bool [@@deriving sexp]

(* Separator between individual sexps. *)
type separator =
  | No_separator
  | Empty_line
[@@deriving sexp]

(* Should closing parentheses be on the same line as the last sexp in the list (modulo
   comments), or should they be on new lines?
   Should opening parentheses always be on the same line as what follows them, or should
   they sometimes (when the first item in the list is a list followed by some other sexp)
   be on a separate line?
*)
type parens =
  | Same_line
  | New_line
[@@deriving sexp]

type t = {
  (* The size of indentation in number of spaces. *)
  indent            :int [@default 2];
  (* Alignment of sexp list into columns. *)
  data_alignment    :data_alignment [@default Data_aligned
                                                (Parens_alignment    false,
                                                 Atom_threshold      6,
                                                 Character_threshold 60,
                                                 Depth_threshold     3)];
  color_scheme      :color array;
  atom_coloring     :atom_coloring [@default Color_first(3)];
  atom_printing     :atom_printing [@default Escaped];
  paren_coloring    :paren_coloring [@default true];
  opening_parens    :parens [@default Same_line];
  closing_parens    :parens [@default Same_line];
  comments          :comments [@default Print(Indent_comment 3,Some Green,Pretty_print)];
  singleton_limit   :
    singleton_limit [@default Singleton_limit(Atom_threshold 3,Character_threshold 15)];
  (* Number of atoms that will be marked as leading in regular lists. They will be put on
     a single line, if they fit.
  *)
  leading_threshold :
    atom_threshold * char_threshold [@default Atom_threshold 3,Character_threshold 20];
  separator         :separator [@default Empty_line];
  (* If true, tries to put comments that correspond to some sexp in front of the sexp
     rather than after it. For example, if the original input is

     SEXP ;comment1
     ;comment2

     then if this is set to true, the output will be

     ;comment1
     ;comment2
     SEXP

     otherwise it will be

     SEXP
     ;comment1
     ;comment2
  *)
  sticky_comments   :bool [@default false];
} [@@deriving sexp]

let t_of_sexp sexp =
  let t = t_of_sexp sexp in
  if t.indent<0
  then of_sexp_error "Indentation must be non-negative." sexp
  else t

let color = function
  | Black   -> 30
  | Red     -> 31
  | Green   -> 32
  | Yellow  -> 33
  | Blue    -> 34
  | Magenta -> 35
  | Cyan    -> 36
  | White   -> 37
  | Default -> 39


src/config.ml
open! Base

let of_sexp_error = Sexplib.Conv.of_sexp_error


type color =
  | Black
  | Red
  | Green
  | Yellow
  | Blue
  | Magenta
  | Cyan
  | White
  | BrightBlack
  | BrightRed
  | BrightGreen
  | BrightYellow
  | BrightBlue
  | BrightMagenta
  | BrightCyan
  | BrightWhite
  | Default
[@@deriving sexp]

(* Datatypes of various thresholds *)
type atom_threshold = Atom_threshold of int [@@deriving sexp]

let atom_threshold_of_sexp sexp =
  match atom_threshold_of_sexp sexp with
  | Atom_threshold n when n < 0 ->
    of_sexp_error "Atom threshold must be non_negative." sexp
  | threshold -> threshold
;;

type char_threshold = Character_threshold of int [@@deriving sexp]

let char_threshold_of_sexp sexp =
  match char_threshold_of_sexp sexp with
  | Character_threshold n when n < 0 ->
    of_sexp_error "Character threshold must be non_negative." sexp
  | threshold -> threshold
;;

(* Depth is the depth of an atom. For example, in (a (b (c) d)), the depth of a is 1, the
   depth of b and d is 2, and depth of c is 3.
   Depth_threshold usually refers to the maximum depth of any atom in a list for it to be
   considered for certain heuristic, e.g. data alignment.
*)
type depth_threshold = Depth_threshold of int [@@deriving sexp]

let depth_threshold_of_sexp sexp =
  match depth_threshold_of_sexp sexp with
  | Depth_threshold n when n < 1 ->
    of_sexp_error "Depth threshold must be greater than 0." sexp
  | threshold -> threshold
;;

(* Whether or not should closing parentheses be aligned. *)
type aligned_parens = Parens_alignment of bool [@@deriving sexp]

type data_alignment =
  | Data_not_aligned
  (* Character threshold is excluding spaces and parentheses, the maximum depth can't exceed
     the depth threshold.
  *)
  | Data_aligned of aligned_parens * atom_threshold * char_threshold * depth_threshold
[@@deriving sexp]

type atom_coloring =
  (* Color the first one, only if the number of atoms that follow it at most the value of
     the constructor's argument.
  *)
  | Color_first of int
  | Color_all
  | Color_none
[@@deriving sexp]

let atom_coloring_of_sexp sexp =
  match atom_coloring_of_sexp sexp with
  | Color_first n when n < 0 ->
    of_sexp_error "The limit to color atoms must be non-negative." sexp
  | coloring -> coloring
;;

(* This currently relates only to block comments. [Auto_indent] tries to infer the
   indentation from the original formatting, [Indent_comment n] indents new lines in a
   block comment by n spaces.
*)
type comment_indent =
  | Auto_indent_comment
  | Indent_comment of int
[@@deriving sexp]

let comment_indent_of_sexp sexp =
  match comment_indent_of_sexp sexp with
  | Indent_comment n when n < 0 -> of_sexp_error "Indentation must be non-negative." sexp
  | indent -> indent
;;

type comment_print_style =
  (* Auto aligns multi-line block comments. *)
  | Pretty_print
  (* Leaves block comments as they are, only adjusts indentation. *)
  | Conservative_print
[@@deriving enumerate, sexp]

(* Comment treatment. *)
type comments =
  | Drop
  | Print of comment_indent * color option * comment_print_style
[@@deriving sexp]

type atom_printing =
  | Escaped (* Can be parsed again. Atoms are printed out as loaded, with escaping. *)
  | Minimal_escaping (* As [Escaped], but applies escaping to fewer characters. *)
  | Interpreted (* Try to interpret atoms as sexps. *)
[@@deriving sexp]

(* Singleton_lists are lists of the following format

   (ATOM_1 .. ATOM_N (....))

   and are printed in the following way if they are too big to fit on a line/force a
   breakline for other reasons:

   (ATOM_1 .. ATOM_N (
   ....
   ))

   Thresholds correspond to what's an acceptable number/size of the leading atoms ATOM_1
   through ATOM_N.

   Character threshold is excluding spaces.

*)
type singleton_limit = Singleton_limit of atom_threshold * char_threshold
[@@deriving sexp]

(* Should parentheses be colored? *)
type paren_coloring = bool [@@deriving sexp]

(* Separator between individual sexps. *)
type separator =
  | No_separator
  | Empty_line
[@@deriving sexp]

(* Should closing parentheses be on the same line as the last sexp in the list (modulo
   comments), or should they be on new lines?
   Should opening parentheses always be on the same line as what follows them, or should
   they sometimes (when the first item in the list is a list followed by some other sexp)
   be on a separate line?
*)
type parens =
  | Same_line
  | New_line
[@@deriving sexp]

(* Where to put the line comments corresponding to some sexp?
   For example, if the original input is

   {v
      SEXP ;comment1
           ;comment2
   v}

   If [Before], put comments in lines before the sexp:

   {v
      ;comment1
      ;comment2
      SEXP
   v}

   If [Same_line], put the first comment right after the sexp, on the same line,
   and align the rest of the comments:

   {v
      SEXP ;comment1
           ;comment2
   v}

   If [After], put comments in lines after the sexp:

   {v
      SEXP
      ;comment1
      ;comment2
   v}
*)
type sticky_comments =
  | Before
  | Same_line
  | After
[@@deriving sexp]

type t =
  { (* The size of indentation in number of spaces. *)
    indent : int [@default 2]
  ; (* Alignment of sexp list into columns. *)
    data_alignment : data_alignment
                     [@default
                       Data_aligned
                         ( Parens_alignment false
                         , Atom_threshold 6
                         , Character_threshold 60
                         , Depth_threshold 3 )]
  ; color_scheme : color array
  ; atom_coloring : atom_coloring [@default Color_first 3]
  ; atom_printing : atom_printing [@default Escaped]
  ; paren_coloring : paren_coloring [@default true]
  ; opening_parens : parens [@default Same_line]
  ; closing_parens : parens [@default Same_line]
  ; comments : comments [@default Print (Indent_comment 3, Some Green, Pretty_print)]
  ; singleton_limit : singleton_limit
                      [@default Singleton_limit (Atom_threshold 3, Character_threshold 15)]
  ; (* Number of atoms that will be marked as leading in regular lists. They will be put on
       a single line, if they fit.
    *)
    leading_threshold : atom_threshold * char_threshold
                        [@default Atom_threshold 3, Character_threshold 20]
  ; separator : separator [@default Empty_line]
  ; sticky_comments : sticky_comments [@default After]
  }
[@@deriving sexp]

let t_of_sexp sexp =
  let t = t_of_sexp sexp in
  if t.indent < 0 then of_sexp_error "Indentation must be non-negative." sexp else t
;;


let default_color_scheme = [| Magenta; Yellow; Cyan; White |]

let default =
  { indent = 2
  ; data_alignment =
      Data_aligned
        ( Parens_alignment false
        , Atom_threshold 6
        , Character_threshold 50
        , Depth_threshold 3 )
  ; color_scheme = default_color_scheme
  ; atom_coloring = Color_first 3
  ; atom_printing = Escaped
  ; paren_coloring = true
  ; closing_parens = Same_line
  ; opening_parens = Same_line
  ; comments = Print (Indent_comment 3, Some Green, Pretty_print)
  ; singleton_limit = Singleton_limit (Atom_threshold 3, Character_threshold 40)
  ; leading_threshold = Atom_threshold 3, Character_threshold 40
  ; separator = Empty_line
  ; sticky_comments = After
  }
;;

let update
      ?color
      ?interpret_atom_as_sexp
      ?drop_comments
      ?new_line_separator
      ?custom_data_alignment
      conf
  =
  let conf =
    match color with
    | None -> conf
    | Some color ->
      if color
      then conf
      else (
        match conf.comments with
        | Print (indent, Some _, style) ->
          { conf with
            atom_coloring = Color_none
          ; paren_coloring = false
          ; comments = Print (indent, None, style)
          }
        | _ -> { conf with atom_coloring = Color_none; paren_coloring = false })
  in
  let conf =
    match interpret_atom_as_sexp with
    | None -> conf
    | Some interpret_atom_as_sexp ->
      if interpret_atom_as_sexp then { conf with atom_printing = Interpreted } else conf
  in
  let conf =
    match drop_comments with
    | None -> conf
    | Some drop_comments -> if drop_comments then { conf with comments = Drop } else conf
  in
  let conf =
    match new_line_separator with
    | None -> conf
    | Some true -> { conf with separator = Empty_line }
    | Some false -> { conf with separator = No_separator }
  in
  let conf =
    match custom_data_alignment with
    | None -> conf
    | Some data_alignment -> { conf with data_alignment }
  in
  conf
;;

let create
      ?(color = false)
      ?(interpret_atom_as_sexp = false)
      ?(drop_comments = false)
      ?(new_line_separator = false)
      ?custom_data_alignment
      ()
  =
  update
    ~color
    ~interpret_atom_as_sexp
    ~drop_comments
    ~new_line_separator
    ?custom_data_alignment
    default
;;
